address.js

```diff
@@ -0,0 +1,263 @@
+const base58 = require('./utils/base58')
+const CborIndefiniteLengthArray = require('./utils/CborIndefiniteLengthArray')
+
+const AddressTypes = {
+  'BASE': 0b0000,
+  'POINTER': 0b0100,
+  'ENTERPRISE': 0b0110,
+  'BOOTSTRAP': 0b1000,
+  'REWARDS': 0b1110
+}
+
+function validateAddressType(input, addressTypeId) {
+  if (!['BASE', 'POINTER', 'ENTERPRISE', 'BOOTSTRAP', 'REWARDS'].includes(addressTypeId)) {
+    throw new Error("Invalid address type to validate!")
+  }
+  const addressType = {
+    BASE: {
+      BASE_ADDRESS_KEY_KEY: 0,
+      BASE_ADDRESS_SCRIPT_KEY: 1,
+      BASE_ADDRESS_KEY_SCRIPT: 2,
+      BASE_ADDRESS_SCRIPT_SCRIPT: 3
+    },
+    POINTER: {
+      POINTER_ADDRESS_KEY: 4,
+      POINTER_ADDRESS_SCRIPT: 5
+    },
+    ENTERPRISE: {
+      ENTERPRISE_ADDRESS_KEY: 6,
+      ENTERPRISE_ADDRESS_SCRIPT: 7
+    },
+    BOOTSTRAP: {
+      BOOTSTRAP_ADDRESS: 8
+    },
+    REWARDS: {
+      REWARDS_ADDRESS: 14
+    }
+  }
+  if (!Object.values(addressType[addressTypeId]).includes(input)) {
+    throw new Error('Invalid address type for address: ' + input
+      + ' Expected one of ' + addressType[addressTypeId])
+  }
+}
+
+function validatePointer(input) {
+  if (!input.hasOwnProperty('blockIndex')
+    || !input.hasOwnProperty('txIndex')
+    || !input.hasOwnProperty('certificateIndex')) {
+    throw new Error('Invalid pointer! Missing one of blockIndex, txIndex, certificateIndex')
+  }
+  if (!Number.isInteger(input.blockIndex)
+    || !Number.isInteger(input.txIndex)
+    || !Number.isInteger(input.certificateIndex)) {
+    throw new Error('Invalid pointer! values must be integer')
+  }
+}
+
+function packBootstrapAddress(derivationPath, xpub, hdPassphrase, derivationScheme) {
+  validateBuffer(xpub, 64)
+  validateDerivationScheme(derivationScheme)
+
+  if (derivationScheme === 1) {
+    validateArray(derivationPath)
+    validateBuffer(hdPassphrase, 32)
+  }
+
+  let addressPayload, addressAttributes
+  if (derivationScheme === 1 && derivationPath.length > 0) {
+    addressPayload = encryptDerivationPath(derivationPath, hdPassphrase)
+    addressAttributes = new Map([[1, cbor.encode(addressPayload)]])
+  } else {
+    addressPayload = Buffer.from([])
+    addressAttributes = new Map()
+  }
+
+  const addressRoot = getAddressHash([
+    0,
+    [0, xpub],
+    addressPayload.length > 0 ? new Map([[1, cbor.encode(addressPayload)]]) : new Map(),
+  ])
+  const addressType = 0 // Public key address
+  const addressData = [addressRoot, addressAttributes, addressType]
+  const addressDataEncoded = cbor.encode(addressData)
+
+  let enc = cbor.encode([new cbor.Tagged(24, addressDataEncoded), crc32(addressDataEncoded)])
+  return enc
+}
+
+function getAddressHeader(addressType, networkId) {
+  return Buffer.from([(addressType << 4) | networkId])
+}
+
+function getSpendingPart(pubKey) {
+  return blake2b(pubKey, 28)
+}
+
+function packBaseAddress(pubKey, stakePubKey, addressType, networkId, isStakeHash = false) {
```

Comment:
stakePubKey possibly meaning stakeHash makes for a rather misleading function signature - we should probably rename the variable to stakePubKeyOrHash and include validateBuffer for the corresponding length to prevent mistakingly injecting the hash instead of the pubkey or viceversa